src/analysis/score.rs

```diff
@@ -24,6 +49,153 @@ pub fn row_accuracy(outputs: &Matrix<f64>, targets: &Matrix<f64>) -> f64 {
+/// ```
+///
+/// # Panics
+///
+/// - outputs and targets have different length
+pub fn precision<'a, I, T>(outputs: I, targets: I) -> f64
+    where I: ExactSizeIterator<Item=&'a T>,
+          T: 'a + PartialEq + Zero + One
+{
+    assert!(outputs.len() == targets.len(), "outputs and targets must have the same length");
+
+    let mut tpfp = 0.0f64;
+    let mut tp = 0.0f64;
+
+    for (ref o, ref t) in outputs.zip(targets) {
+        if *o == &T::one() {
+            tpfp += 1.0f64;
+            if *t == &T::one() {
+                tp += 1.0f64;
+            }
+        }
+        if ((*t != &T::zero()) & (*t != &T::one())) |
+           ((*o != &T::zero()) & (*o != &T::one())) {
+            panic!("precision must be used for 2 class classification")
+        }
+    }
+    tp / tpfp
+}
+
+/// Returns the recall score for 2 class classification.
+///
+/// Recall is calculated with true-positive / (true-positive + false-negative),
+/// see [Precision and Recall](https://en.wikipedia.org/wiki/Precision_and_recall) for details.
+///
+/// # Arguments
+///
+/// * `outputs` - Iterator of output (predicted) labels which only contains 0 or 1.
+/// * `targets` - Iterator of expected (actual) labels which only contains 0 or 1.
+///
+/// # Examples
+///
+/// ```
+/// use rusty_machine::analysis::score::recall;
+/// let outputs = [1, 1, 1, 0, 0, 0];
+/// let targets = [1, 1, 0, 0, 1, 1];
+///
+/// assert_eq!(recall(outputs.iter(), targets.iter()), 0.5);
+/// ```
+///
+/// # Panics
+///
+/// - outputs and targets have different length
+pub fn recall<'a, I, T>(outputs: I, targets: I) -> f64
+    where I: ExactSizeIterator<Item=&'a T>,
+          T: 'a + PartialEq + Zero + One
+{
+    assert!(outputs.len() == targets.len(), "outputs and targets must have the same length");
+
+    let mut tpfn = 0.0f64;
+    let mut tp = 0.0f64;
+
+    for (ref o, ref t) in outputs.zip(targets) {
+        if *t == &T::one() {
+            tpfn += 1.0f64;
+            if *o == &T::one() {
+                tp += 1.0f64;
+            }
+        }
+        if ((*t != &T::zero()) & (*t != &T::one())) |
+           ((*o != &T::zero()) & (*o != &T::one())) {
+            panic!("recall must be used for 2 class classification")
+        }
+    }
+    tp / tpfn
+}
+
+/// Returns the f1 score for 2 class classification.
+///
+/// F1-score is calculated with 2 * precision * recall / (precision + recall),
+/// see [F1 score](https://en.wikipedia.org/wiki/F1_score) for details.
+///
+/// # Arguments
+///
+/// * `outputs` - Iterator of output (predicted) labels which only contains 0 or 1.
+/// * `targets` - Iterator of expected (actual) labels which only contains 0 or 1.
+///
+/// # Examples
+///
+/// ```
+/// use rusty_machine::analysis::score::f1;
+/// let outputs = [1, 1, 1, 0, 0, 0];
+/// let targets = [1, 1, 0, 0, 1, 1];
+///
+/// assert_eq!(f1(outputs.iter(), targets.iter()), 0.5714285714285714);
+/// ```
+///
+/// # Panics
+///
+/// - outputs and targets have different length
```

Comment:
I'm not sure why all my comments got placed here.. This is certainly not where I left them...